credoai/evaluators/shap.py

```diff
@@ -0,0 +1,141 @@
+from numpy import mean, abs
+
+
+class ShapValues(Evaluator):
+    """
+    This evaluator perform the calculation of shapley values for a dataset/model,
+    leveraging the SHAP package.
+
+    It calculates 2 types of assessments:
+    1. Overall statistics of the shap values across all samples: mean and mean(|x|)
+    2. Optional: individual shapley values for a list of samples
+
+    Parameters
+    ----------
+    samples_ind : Optional[List[int]], optional
+        List of row numbers representing the samples for which to extract individual
+        shapley values , by default None
+    """
+
+    def __init__(self, samples_ind: Optional[List[int]] = None):
+
+        super().__init__()
+        self.samples_ind = samples_ind
+
+    name = "Shap"
+    required_artifacts = ["assessment_data", "model"]
+
+    def _setup(self):
+        self.X = self.assessment_data.X
+        self.model = self.model
+        return self
+
+    def _validate_arguments(self):
+        check_requirements_existence(self)
+
+    def evaluate(self):
+        ## Overall stats
+        self._setup_shap()
+        res = self._get_overall_shap_contributions()
+        self.results = [TableContainer(res, **self.get_container_info())]
+
+        ## Sample specific results
+        if self.samples_ind:
+            labels = {"ordered_feature_names": self.shap_values.feature_names}
+            ind_res = self._get_mult_sample_shapley_values()
+            self.results += [
+                TableContainer(ind_res, **self.get_container_info(labels=labels))
+            ]
+        return self
+
+    def _setup_shap(self):
+        """
+        Setup the explainer given the model and the feature dataset
+        """
+        explainer = Explainer(self.model.predict, self.X)
+        self.shap_values = explainer(self.X)
+        return self
+
+    def _get_overall_shap_contributions(self) -> DataFrame:
+        """
+        Calculate overall SHAP contributions for a dataset.
+
+        The output of SHAP package provides Shapley values for each sample in a
+        dataset. To summarise the contribution of each feature in a dataset, the
+        samples contributions need to be aggregated.
+
+        For each of the features, this method provides: mean, max, minimum and the
+        mean of the absolute value of the samples shapley values.
+
+        Returns
+        -------
+        DataFrame
+            Summary of the shapley values across the full dataset.
+        """
+        values_df = DataFrame(self.shap_values.values)
+        values_df.columns = self.shap_values.feature_names
+        stats_1 = values_df.apply(["mean"]).T
+        stats_2 = values_df.apply(lambda x: mean(abs(x)))
+        stats_2.name = "mean(|x|)"
+        final = concat([stats_1, stats_2], axis=1)
+        final = final.sort_values("mean(|x|)", ascending=False)
+        final.name = "Summary of Shap statistics"
+
+        return final
+
+    def _get_mult_sample_shapley_values(self) -> DataFrame:
```

Comment:
One major concern about this is that it could reveal too much information about the underlying dataset nor the internals of the model.
For instance, we don't want someone to generate a report that has a full sample-by-sample output of how the model made it's predictions. It'd be a very long report so I think it's unlikely that someone would do that. I also don't know exactly how we can arbitrate what people do/don't try (is 5 individual samples too many? 10?).

Something to think about...

Another thought is whether it makes sense to intersect this (and how to do it) with groupby functionalities we've implemented elsewhere. For instance, maybe it makes more sense to allow a report to be generated with SHAP values based on males/females separately than it does to worry about individual-sample-SHAP

^ could be either a) mean/abs_mean shap values calculated on an all-male or all-female dataset or b) calculating shap values on a combined dataset and then reporting means from subsets (not sure off-hand if those will be mathematically guaranteed to be the same or different...)